proposals/roles.md

```diff
@@ -0,0 +1,155 @@
+﻿# Roles
+
+This proposal is divided into three parts, all relating to extending existing types:
+A. static extensions
+B. roles and extensions with members
+C. roles and extensions that implement interfaces
+
+The syntax for a roles is as follows:
+
+```antr
+type_declaration
+	: role_declaration // add
+	| extension_declaration // add
+	| ...
+	;
+
+role_declaration
+    : modifier* 'role' identifier type_parameter_list? role_underlying_type type_parameter_constraints_clause* role_body
+    ;
+
+extension_declaration
+    : modifier* 'extension' identifier type_parameter_list? role_underlying_type type_parameter_constraints_clause* role_body
+    ;
+
+role_underlying_type
+    : ':' type
+    ;
+
+role_body
+    : '{' role_member_declaration* '}'
+    ;
+
+role_member_declaration
+    : constant_declaration
+    | field_declaration
+    | method_declaration
+    | property_declaration
+    | event_declaration
+    | indexer_declaration
+    | operator_declaration
+    | type_declaration
+    ;
+```
+
+## Phase A: Adding static constants, fields, methods and properties
+
+In this first subset of the feature, the syntax is restricted to non-nested **extension_declaration**
+and containing only **constant_declaration**, **field_declaration**, **method_declaration** and **property_declaration** members.
+TODO: events?
+
+### General
+
+The permitted modifiers are `partial`, `static`, `unsafe` and the accessibility modifiers `public` and `internal`.  
+An extension declaration must have a `static` modifier. If the declaration is partial, then at least one part must have the `static` modifier. 
+The standard rules for modifiers apply (valid combination of access modifiers, no duplicates).
+The **role_underlying_type** type may not be `dynamic`, a pointer or an extension type.  
+This declares a static class type. It inherits from type `object`. It has `internal` declared accessibility by default.
+TODO: attributes?
+TODO: emitted name?
+
+### Members
+
+The extension type members may not use the `new` modifier.
+Accessibility modifiers including `protected` are disallowed.
```

Comment:
oh, interesting.  seems a pity as i could imagine having internal members that are useful for when looking through the role in its own assembly, versus public members when seen externally.